File: src/main/java/com/oppshan/washa/exception/MessageCode.java
package com.oppshan.washa.exception;

import com.fasterxml.jackson.annotation.JsonValue;

/**
 * The backend's half of the {@code MessageCode} contract shared with the frontend: the TS enum
 * ({@code message-code.ts}) and the i18n keys in {@code en.json}. Each constant's string IS an
 * en.json key path; Jackson emits it via {@link JsonValue}, so that key is what ships in an error body
 * or a redirect query and the frontend resolves it to a localized message. TS is the superset (it adds
 * frontend-only info and validation codes); this enum carries just the codes the backend side needs.
 *
 * <p>Most codes are emitted by the backend. {@code UNKNOWN} (the frontend's fallback for an
 * unrecognized code) and {@code SIGN_IN_FAILED} aren't produced by any current backend path; they
 * exist here only to stay byte-for-byte aligned with the TS enum. Change a value in lockstep across
 * Java, TS, and en.json, or the wire contract drifts (root CLAUDE.md C.1).
 */
public enum MessageCode {
    UNKNOWN("messages.errors.unknown"),
    AUTHENTICATION_REQUIRED("messages.errors.authenticationRequired"),
    ACCESS_DENIED("messages.errors.accessDenied"),
    SIGN_IN_FAILED("messages.errors.signInFailed"),
    USER_NOT_FOUND("messages.errors.userNotFound"),
    SALARY_PRESET_NOT_FOUND("messages.errors.salaryPresetNotFound"),
    SALARY_PRESET_BUILT_IN("messages.errors.salaryPresetBuiltIn");

    private final String key;

    /** Binds each constant to its en.json key path. */
    MessageCode(String key) {
        this.key = key;
    }

    /** The en.json key path; Jackson serializes this as the code's wire value via {@code @JsonValue}. */
    @JsonValue
    public String getKey() {
        return key;
    }
}

File: src/main/angular/src/app/models/message-code.ts
/**
 * The frontend's half of the `MessageCode` contract (root CLAUDE.md C.1). Each value is an i18n key
 * path resolved against `en.json`, and severity comes from the prefix (`messages.errors.*` = error,
 * `messages.info.*` = info). This enum is the superset: it mirrors the Java `MessageCode` byte-for-byte
 * for every code the backend emits, then adds frontend-only codes that never cross the wire.
 */
export enum MessageCode {
  /**
   * Backend-mirrored codes: every value from here through `SalaryPresetBuiltIn` is byte-equal with a
   * Java `MessageCode` constant, so the wire contract holds. `Unknown` (the fallback for an
   * unrecognized code) and `SignInFailed` keep Java counterparts for alignment even though no current
   * backend path emits them.
   */
  Unknown = 'messages.errors.unknown',
  AuthenticationRequired = 'messages.errors.authenticationRequired',
  AccessDenied = 'messages.errors.accessDenied',
  SignInFailed = 'messages.errors.signInFailed',
  UserNotFound = 'messages.errors.userNotFound',
  SalaryPresetNotFound = 'messages.errors.salaryPresetNotFound',
  SalaryPresetBuiltIn = 'messages.errors.salaryPresetBuiltIn',

  /**
   * Frontend-only success and lifecycle codes; the backend never emits these, so Java has no
   * counterpart (a Java entry would be a dead enum value).
   */
  SignInSucceeded = 'messages.info.signInSucceeded',
  BudgetSaved = 'messages.info.budgetSaved',
  MonthLoaded = 'messages.info.monthLoaded',
  ImportSucceeded = 'messages.info.importSucceeded',
  FxRefreshed = 'messages.info.fxRefreshed',

  /** Frontend-only validation code (error severity), raised client-side before any request goes out. */
  ImportRejected = 'messages.errors.importRejected',
}

/**
 * Resolves a raw wire string to its `MessageCode`, falling back to `Unknown` when the value is
 * missing or unrecognized, so a code that has drifted out of sync degrades gracefully rather than
 * throwing.
 */
export function messageCodeOf(value: string | null | undefined): MessageCode {
  if (!value) {
    return MessageCode.Unknown;
  }
  const match = Object.values(MessageCode).find((code) => code === value);
  return (match as MessageCode) ?? MessageCode.Unknown;
}
